src/torchjd/autogram/_module_hook_manager.py

```diff
@@ -0,0 +1,129 @@
+from typing import cast
+
+import torch
+from torch import Tensor, nn
+from torch.autograd.graph import get_gradient_edge
+from torch.utils._pytree import PyTree, TreeSpec, tree_flatten, tree_unflatten
+from torch.utils.hooks import RemovableHandle as TorchRemovableHandle
+
+from ._edge_registry import EdgeRegistry
+from ._gramian_accumulator import GramianAccumulator
+from ._vjp import get_functional_vjp
+
+# Note about import from protected _pytree module:
+# PyTorch maintainers plan to make pytree public (see
+# https://github.com/pytorch/pytorch/issues/65761, https://github.com/pytorch/pytorch/pull/137400).
+# It should also come with better speed, because the current implementation is slow, according to
+# https://github.com/pytorch/pytorch/issues/65761#issue-1010116111.
+# When pytree becomes public, this import will have to be changed with a conditional import (to
+# still support older versions of PyTorch where pytree is protected).
+
+
+class ModuleHookManager:
+    """
+    Class responsible for handling hooks and Nodes that computes the Gramian reverse accumulation.
+
+    :param target_edges: Registry for tracking gradient edges that serve as targets for the first
+        differentiation.
+    :param gramian_accumulator: Accumulator for collecting the Jacobians into a Gramian.
+    """
+
+    def __init__(
+        self,
+        target_edges: EdgeRegistry,
+        gramian_accumulator: GramianAccumulator,
+    ):
+        self._target_edges = target_edges
+        self._gramian_accumulator = gramian_accumulator
+        self.gramian_accumulation_phase = False
+        self._handles: list[TorchRemovableHandle] = []
+
+    def hook_module(self, module: nn.Module) -> None:
+        """
+        Add a module hook used to insert Jacobian accumulation nodes into the backward graph.
+
+        The hook injects a JacobianAccumulator function into the computation graph after the module,
+        enabling Gramian computation during autogram's first backward pass.
```

Comment:
```suggestion
        enabling Gramian computation.
```

There is no more 2 phases, it is during the Gramian accumulation phase, but this is tautological.